### PR TITLE
feat(LERN-68): user can now create grades

### PR DESCRIPTION
**Pull request by kieranslost**

### CODE REVIEW

File: frontend/src/pages/GradePage.tsx [hunk 1]
```diff
@@ -1,45 +1,111 @@
...
 import { AlternateButton } from '../components/AlternateButton.tsx';
 import { ContentSection } from '../components/ContentSection.tsx';
 import { GradeCard } from '../components/GradeCard.tsx';
-import { useGetGradesForSchoolQuery } from '../state/api/grades.ts';
+import {useGetGradesForSchoolQuery} from '../state/api/grades.ts';
 import { whoamiAtom } from '../state/api/whoami.ts';
+import {useMutation} from "@tanstack/react-query";
+import {
+    AdminGradeResourceService,
+    type CreateGradeDto,
```
Comment (deepspace1000) [inline frontend/src/pages/GradePage.tsx:12]: I guess you've don that manually? you should not need write the imports manually. if you do you probably did something wrong. 
I'm talking about the type import, i've never seen that 🤔

**Review by deepspace1000 — CHANGES_REQUESTED** [review]
Just had a short look.
move the modal to its own file and afterwards i will give you a propper review. also run the formatter 🙃

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 0       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 1  | ❗       |
    


*❗ custom/admin-landing-page*: 


| Before | After |
| --- | --- |
|  |  |

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 0       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 1  | ❗       |
    


*❗ custom/admin-landing-page*: 


| Before | After |
| --- | --- |
|  |  |

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 0       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 1  | ❗       |
    


*❗ custom/admin-landing-page*: 


| Before | After |
| --- | --- |
|  |  |

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 0       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 1  | ❗       |
    


*❗ custom/admin-landing-page*: 


| Before | After |
| --- | --- |
|  |  |

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 0       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 1  | ❗       |
    


*❗ custom/admin-landing-page*: 


| Before | After |
| --- | --- |
|  |  |

File: frontend/src/components/CreateGrade.tsx [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  return (
+    <Modal open={isOpen} onClose={handleModalClose}>
+      <ModalDialog>
+        <Typography level="h4">Neue Klasse erstellen</Typography>
+        <Input
+          value={gradeName}
+          onChange={(e) => setGradeName(e.target.value)}
+          placeholder="Geben sie den namen ein"
+          fullWidth
+        />
```
Comment (deepspace1000) [inline frontend/src/components/CreateGrade.tsx:48]: I would probaly still go with react hook forms, even thought its arguable if it makes sense for only one input. But it does make validation and error handling easier... So what would you say?

File: frontend/src/components/CreateGrade.tsx [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        <div style={{ display: 'flex', justifyContent: 'flex-end', gap: '8px' }}>
+          <Button onClick={handleModalClose}>Cancel</Button>
```
Comment (deepspace1000) [inline frontend/src/components/CreateGrade.tsx:50]: just call the on close function directly.

File: frontend/src/components/CreateGrade.tsx [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    if (gradeName != '') {
+      createGrade.mutate({ name: gradeName, schoolId });
+    } else {
+      setShowSnackbarError(true);
+    }
+    setOpen(false);
+  };
+
+  return (
+    <Modal open={isOpen} onClose={handleModalClose}>
```
Comment (deepspace1000) [inline frontend/src/components/CreateGrade.tsx:40]: call the onClose function directly

File: frontend/src/pages/GradePage.tsx [hunk 5]
```diff
@@ -1,24 +1,47 @@
...
+  const schoolId = whoami.schools[0].id;
+  const [open, setOpen] = useState(false);
+
+  const grades = useGetGradesForSchoolQuery(schoolId);
+  const queryClient = useQueryClient();
+
+  const [showSnackbar, setShowSnackbar] = useState(false);
+  const [showSnackbarError, setShowSnackbarError] = useState(false);
+
+  useEffect(() => {
```
Comment (deepspace1000) [inline frontend/src/pages/GradePage.tsx:24]: Wrong place 😂. the invalidate query should be called with the useMutation onSuccess.

Comment (deepspace1000) [reply frontend/src/pages/GradePage.tsx:24]: Also for the build fail a little tip:
```
Promise.all([
        queryClient.invalidateQueries({
          queryKey: ['gradeList', shoolId],
        }),
```

File: frontend/src/pages/GradePage.tsx [hunk 6]
```diff
@@ -31,15 +54,34 @@ export function GradePage() {
...
+      <Snackbar
+        open={showSnackbarError}
+        color={'danger'}
+        autoHideDuration={3000}
+        onClose={() => {
+          setShowSnackbarError(false);
+        }}
+        startDecorator={<ErrorOutline />}>
+        Ein Fehler ist wärend der Kreierung aufgetreten
+      </Snackbar>
```
Comment (deepspace1000) [inline frontend/src/pages/GradePage.tsx:75]: Move those to the Modal file too

**Review by deepspace1000 — COMMENTED** [review]
Some stuff you should have a look at 😉. If you have questions just ask me.

File: frontend/src/components/CreateGrade.tsx [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+import { useAtom } from 'jotai/index';
+import { useState } from 'react';
+import { AdminGradeResourceService, CreateGradeDto } from '../state/api/generated';
+import { whoamiAtom } from '../state/api/whoami.ts';
+
+export function CreateGrade({
+  isOpen,
+  setOpen,
+  setShowSnackbar,
+  setShowSnackbarError,
```
[the remark this answers is not in the document]
Comment (kieranslost) [reply frontend/src/components/CreateGrade.tsx:12]: ez, i'll try

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 0       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 1  | ❗       |
    


*❗ custom/admin-landing-page*: 


| Before | After |
| --- | --- |
|  |  |

File: frontend/src/components/CreateGradeDialog.tsx [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+  const queryClient = useQueryClient();
+
+  const [showSnackbar, setShowSnackbar] = useState(false);
+  const [showSnackbarError, setShowSnackbarError] = useState(false);
+
+  const createGrade = useCreateGradeForSchoolMutation(
+    schoolId,
+    () => {
+      setShowSnackbar(true);
+      queryClient.invalidateQueries(['gradeList', schoolId]);
```
Comment (deepspace1000) [inline frontend/src/components/CreateGradeDialog.tsx:22]: Put the invalidate into the useMutation directly

File: frontend/src/components/CreateGradeDialog.tsx [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+      </Snackbar>
+      <Snackbar
+        open={showSnackbarError}
+        color={'danger'}
+        autoHideDuration={3000}
+        onClose={() => {
+          setShowSnackbarError(false);
+        }}
+        startDecorator={<ErrorOutline />}>
+        Ein Fehler ist wärend der Kreierung aufgetreten
```
Comment (deepspace1000) [inline frontend/src/components/CreateGradeDialog.tsx:70]: Beim erstellend der Klasse ist ein Fehler aufgetreten.

File: frontend/src/pages/GradePage.tsx [hunk 10]
```diff
@@ -1,24 +1,31 @@
...
+  const grades = useGetGradesForSchoolQuery(schoolId);
 
   return (
     <ContentSection>
       <div style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', marginBottom: '40px' }}>
         <Typography level="h3" margin={0} lineHeight={1}>
           Klassenübersicht
         </Typography>
-        <AlternateButton>Klasse hinzufügen</AlternateButton>
+        <AlternateButton type="submit" onClick={() => setOpen(!open)}>
```
Comment (deepspace1000) [inline frontend/src/pages/GradePage.tsx:24]: You don't need type submit as you don't submit a Form

File: frontend/src/pages/GradePage.tsx [hunk 11]
```diff
@@ -36,10 +43,10 @@ export function GradePage() {
 }
 
 const GradesCardWrapper = styled('div')`
-  display: flex;
-  flex-wrap: wrap;
+  display: grid;
+  grid-template-columns: repeat(3, auto);
   align-items: center;
-  justify-content: space-between;
+  justify-items: center;
```
Comment (deepspace1000) [inline frontend/src/pages/GradePage.tsx:49]: I think i had a look at it with martin and we decided to leave it for now as this approach comes with its own problems.
@Quintexxx right?

**Review by deepspace1000 — APPROVED** [review]

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 0       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 0  | ❗       |

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 0       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 0  | ❗       |

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 0       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 1  | ❗       |
    


*❗ custom/admin-landing-page*: 


| Before | After |
| --- | --- |
|  |  |

**Pull request merged**